pkg/protocol/block_dispatcher.go

```diff
@@ -205,21 +214,131 @@ func (b *BlockDispatcher) processWarpSyncResponse(commitmentID iotago.Commitment
+	}
+
+	if !iotago.VerifyProof(mutationMerkleProof, iotago.Identifier(acceptedTransactionIDs.Root()), chainCommitment.Commitment().RootsID()) {
+		return ierrors.Errorf("failed to verify mutation merkle proof for %s", commitmentID)
 	}
 
 	b.pendingWarpSyncRequests.StopTicker(commitmentID)
 
 	b.processedWarpSyncRequests.Add(commitmentID)
 
+	// If the engine is "dirty" we need to restore the state of the engine to the state of the chain commitment.
+	// As we already decided to switch and sync to this chain we should make sure that processing the blocks from the commitment
+	// leads to the verified commitment.
+	if targetEngine.Notarization.AcceptedBlocksCount(commitmentID.Slot()) > 0 {
+		b.protocol.activeEngineMutex.Lock()
+
+		newEngine, err := b.protocol.EngineManager.RollbackEngine(commitmentID.Slot() - 1)
+		if err != nil {
+			return ierrors.Wrapf(err, "failed to rollback engine for slot %d", commitmentID.Slot())
+		}
+		b.protocol.Events.MainEngineRestarted.Trigger(newEngine)
+
+		newEngine.SetChainID(targetEngine.ChainID())
+
+		if err := b.protocol.EngineManager.SetActiveInstance(newEngine); err != nil {
+			return ierrors.Wrap(err, "failed to set active engine instance")
+		}
+
+		b.protocol.linkToEngine(newEngine)
+
+		b.protocol.mainEngine.Shutdown()
+		b.protocol.mainEngine = newEngine
+		targetEngine = newEngine
+		b.protocol.activeEngineMutex.Unlock()
+	}
+
+	// Once all blocks are booked we
+	//   1. Mark all transactions as accepted
+	//   2. Mark all blocks as accepted
+	//   3. Force commitment of the slot
+	totalBlocks := uint32(len(blockIDs))
+	var bookedBlocks atomic.Uint32
+	var notarizedBlocks atomic.Uint32
+
+	forceCommitmentFunc := func() {
+		// 3. Force commitment of the slot
+		producedCommitment, err := targetEngine.Notarization.ForceCommit(commitmentID.Slot())
+		if err != nil {
+			b.protocol.HandleError(err)
+			return
+		}
+
+		// 4. Verify that the produced commitment is the same as the initially requested one
+		if producedCommitment.ID() != commitmentID {
+			b.protocol.HandleError(ierrors.Errorf("producedCommitment ID mismatch: %s != %s", producedCommitment.ID(), commitmentID))
+			return
+		}
+	}
+
+	blockBookedFunc := func(_, _ bool) {
+		if bookedBlocks.Add(1) != totalBlocks {
+			return
+		}
+
+		// 1. Mark all transactions as accepted
+		for _, transactionID := range transactionIDs {
+			targetEngine.Ledger.ConflictDAG().SetAccepted(transactionID)
+		}
+
+		// 2. Mark all blocks as accepted
+		for _, blockID := range blockIDs {
+			block, exists := targetEngine.BlockCache.Block(blockID)
+			if !exists { // this should never happen as we just booked these blocks in this slot.
+				continue
+			}
+
+			targetEngine.BlockGadget.SetAccepted(block)
+
+			block.Notarized().OnUpdate(func(_, _ bool) {
+				// Wait for all blocks to be notarized before forcing the commitment of the slot.
+				if notarizedBlocks.Add(1) != totalBlocks {
+					return
+				}
+
+				forceCommitmentFunc()
```

Comment:
There is a race condition here between concurrent warpsync responses:

```
panic: there is a gap in the ledgerstate 4998 vs 4997

goroutine 138 [running]:
github.com/iotaledger/iota-core/pkg/protocol/engine/ledger/ledger.(*Ledger).CommitSlot(0xc0005caf20, 0x1385)
	/scratch/iota-core/pkg/protocol/engine/ledger/ledger/ledger.go:148 +0xa85
github.com/iotaledger/iota-core/pkg/protocol/engine/notarization/slotnotarization.(*Manager).createCommitment(0xc0001e9140, 0x1385)
	/scratch/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:183 +0x23c
github.com/iotaledger/iota-core/pkg/protocol/engine/notarization/slotnotarization.(*Manager).ForceCommit(0xc0001e9140, 0xe257a30?)
	/scratch/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:108 +0x45
github.com/iotaledger/iota-core/pkg/protocol.(*BlockDispatcher).processWarpSyncResponse.func1()
	/scratch/iota-core/pkg/protocol/block_dispatcher.go:281 +0x6c
github.com/iotaledger/iota-core/pkg/protocol.(*BlockDispatcher).processWarpSyncResponse(0xc0003bf810, {0x6b, 0x93, 0xe1, 0x11, 0x38, 0x2c, 0xb6, 0x1, 0x61, ...}, ...)
	/scratch/iota-core/pkg/protocol/block_dispatcher.go:325 +0xbad
github.com/iotaledger/iota-core/pkg/protocol.(*BlockDispatcher).initNetworkConnection.func5.1()
	/scratch/iota-core/pkg/protocol/block_dispatcher.go:159 +0x7a
github.com/iotaledger/hive.go/runtime/workerpool.(*Task).run(0xc0135c2cc0)
	/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20230929122509-67f34bfed40d/workerpool/task.go:48 +0x72
github.com/iotaledger/hive.go/runtime/workerpool.(*WorkerPool).workerReadLoop(0xc0002221c0)
	/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20230929122509-67f34bfed40d/workerpool/workerpool.go:177 +0x25
github.com/iotaledger/hive.go/runtime/workerpool.(*WorkerPool).worker(0xc0002221c0)
	/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20230929122509-67f34bfed40d/workerpool/workerpool.go:159 +0x50
created by github.com/iotaledger/hive.go/runtime/workerpool.(*WorkerPool).startWorkers in goroutine 1
	/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20230929122509-67f34bfed40d/workerpool/workerpool.go:151 +0x25
```

Reply:
```
panic: there is a gap in the ledgerstate 5152 vs 5151

goroutine 180 [running]:
github.com/iotaledger/iota-core/pkg/protocol/engine/ledger/ledger.(*Ledger).CommitSlot(0xc00040fa20, 0x141f)
	/scratch/iota-core/pkg/protocol/engine/ledger/ledger/ledger.go:148 +0xa85
github.com/iotaledger/iota-core/pkg/protocol/engine/notarization/slotnotarization.(*Manager).createCommitment(0xc0003b40c0, 0x141f)
	/scratch/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:183 +0x23c
github.com/iotaledger/iota-core/pkg/protocol/engine/notarization/slotnotarization.(*Manager).tryCommitSlotUntil(0xc0003b40c0, 0x142a)
	/scratch/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:155 +0x89
github.com/iotaledger/iota-core/pkg/protocol/engine/notarization/slotnotarization.(*Manager).tryCommitUntil(0xc0003b40c0, 0x142a)
	/scratch/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:99 +0x45
github.com/iotaledger/iota-core/components/protocol.provide.func1.NewProvider.func14.1.1(0xc000622ec8?)
	/scratch/iota-core/pkg/protocol/engine/notarization/slotnotarization/manager.go:66 +0x131
github.com/iotaledger/hive.go/runtime/event.(*Event1[...]).Trigger.func1.1()
	/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20230929122509-67f34bfed40d/event/events.go:86 +0x1f
github.com/iotaledger/hive.go/runtime/workerpool.(*Task).run(0xc018935740)
	/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20230929122509-67f34bfed40d/workerpool/task.go:48 +0x72
github.com/iotaledger/hive.go/runtime/workerpool.(*WorkerPool).workerReadLoop(0xc000252cb0)
	/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20230929122509-67f34bfed40d/workerpool/workerpool.go:177 +0x25
github.com/iotaledger/hive.go/runtime/workerpool.(*WorkerPool).worker(0xc000252cb0)
	/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20230929122509-67f34bfed40d/workerpool/workerpool.go:159 +0x50
created by github.com/iotaledger/hive.go/runtime/workerpool.(*WorkerPool).startWorkers in goroutine 1
	/go/pkg/mod/github.com/iotaledger/hive.go/runtime@v0.0.0-20230929122509-67f34bfed40d/workerpool/workerpool.go:151 +0x25
```

Reply:
Fixed by locking notarization's `createCommitment` https://github.com/iotaledger/iota-core/pull/379/commits/b0dd8dc7303dce3653c35808c5d14401c2421236